executor/new_executor.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/evaluator"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+)
+
+// HashJoinExec implements the hash join algorithm
+type HashJoinExec struct {
+	hashTable    map[string][]*Row
+	smallHashKey []ast.ExprNode
+	bigHashKey   []ast.ExprNode
+	smallExec    Executor
+	bigExec      Executor
+	prepared     bool
+	fields       []*ast.ResultField
+	ctx          context.Context
+	smallFilter  ast.ExprNode
+	bigFilter    ast.ExprNode
+	otherFilter  ast.ExprNode
+	outter       bool
+	leftSmall    bool
+}
+
+func joinTwoRow(a *Row, b *Row) *Row {
+	return &Row{RowKeys: append(a.RowKeys, b.RowKeys...), Data: append(a.Data, b.Data...)}
+}
+
+func (e *HashJoinExec) getHashKey(exprs []ast.ExprNode) (string, error) {
+	vals := make([]types.Datum, 0, len(exprs))
+	for _, expr := range exprs {
+		v, err := evaluator.Eval(e.ctx, expr)
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+		vals = append(vals, v)
+	}
+	if len(vals) == 0 {
+		return "SingleHashGroup", nil
+	}
+	result, err := codec.EncodeValue([]byte{}, vals...)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	return string(result), err
```

Comment:
You can try the trick in this slice: https://github.com/gopherchina/conference/blob/master/2016/2.1%20Writing%20High%20Performance%20Go.pdf
using []byte as a map key